src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/TempPE/DesignTimeInputsBuildManagerBridge.cs

```diff
@@ -0,0 +1,141 @@
+{
+    [Export(typeof(IDesignTimeInputsBuildManagerBridge))]
+    [AppliesTo(ProjectCapability.CSharpOrVisualBasicLanguageService)]
+    internal class DesignTimeInputsBuildManagerBridge : UnconfiguredProjectHostBridge<IProjectVersionedValue<DesignTimeInputsDelta>, IProjectVersionedValue<DesignTimeInputsDelta>, IProjectVersionedValue<DesignTimeInputsDelta>>, IDesignTimeInputsBuildManagerBridge
+    {
+        private readonly IProjectThreadingService _threadingService;
+        private readonly IDesignTimeInputsChangeTracker _designTimeInputsChangeTracker;
+        private readonly IDesignTimeInputsCompiler _designTimeInputsCompiler;
+        private readonly VSBuildManager _buildManager;
+
+        /// <summary>
+        /// For unit testing purposes, to avoid having to mock all of CPS
+        /// </summary>
+        internal bool SkipInitialization { get; set; }
+
+        [ImportingConstructor]
+        public DesignTimeInputsBuildManagerBridge(IProjectThreadingService threadingService,
+                                                  IDesignTimeInputsChangeTracker designTimeInputsChangeTracker,
+                                                  IDesignTimeInputsCompiler designTimeInputsCompiler,
+                                                  VSBuildManager buildManager)
+             : base(threadingService.JoinableTaskContext)
+        {
+            _threadingService = threadingService;
+            _designTimeInputsChangeTracker = designTimeInputsChangeTracker;
+            _designTimeInputsCompiler = designTimeInputsCompiler;
+            _buildManager = buildManager;
+        }
+
+        /// <summary>
+        /// Get the list of design time monikers that need to have TempPE libraries created. Needs to be called on the UI thread.
+        /// </summary>
+        public string[] GetTempPEMonikers()
+        {
+            // Not using use the ThreadingService property because unit tests
+            _threadingService.VerifyOnUIThread();
+
+            Initialize();
+
+            return AppliedValue.Value.Inputs.ToArray();
+        }
+
+        /// <summary>
+        /// Gets the XML that describes a TempPE DLL, including building it if necessary
+        /// </summary>
+        public async Task<string> GetDesignTimeInputXmlAsync(string file)
+        {
+            // Not using use the ThreadingService property because unit tests
+            await _threadingService.SwitchToUIThread();
+
+            if (!SkipInitialization)
+            {
+                await InitializeAsync();
+            }
+
+            DesignTimeInputsDelta value = AppliedValue.Value;
+
+            return await _designTimeInputsCompiler.GetDesignTimeInputXmlAsync(file, value.TempPEOutputPath, value.SharedInputs);
+        }
+
+        /// <summary>
+        /// ApplyAsync is called on the UI thread and its job is to update AppliedValue to be correct based on the changes that have come through data flow after being processed
```

Comment:
Is this correct? It says "called on the UI thread", but the first thing it does is switch to the UI thread.

Reply:
There is an analyser that triggers if you use VerifyOnUIThread in an async method